katz/src/main/kotlin/katz/effects/IO.kt

```diff
@@ -0,0 +1,228 @@
+
+    override fun unsafeStep(): IO<A> =
+            cont(Unit).unsafeStep()
+
+    override fun unsafeRunAsyncTotal(cb: (Either<Throwable, A>) -> Unit) =
+            throw AssertionError("Unreachable")
+
+    override fun unsafeRunTimedTotal(limit: Duration): Option<A> =
+            throw AssertionError("Unreachable")
+}
+
+internal data class BindSuspend<E, out A>(val cont: AndThen<Unit, IO<E>>, val f: AndThen<E, IO<A>>) : IO<A>() {
+    override fun <B> map(f: (A) -> B): IO<B> =
+            mapDefault(this, f)
+
+    override fun <B> flatMapTotal(ff: AndThen<A, IO<B>>): IO<B> =
+            BindSuspend(cont, f.andThen(AndThen({ it.flatMapTotal(ff) }, { ff.error(it, { RaiseError(it) }) })))
+
+    override fun attempt(): IO<Either<Throwable, A>> =
+            BindSuspend(AndThen { _ -> this }, attemptValue())
+
+    override fun unsafeStep(): IO<A> =
+            cont(Unit).flatMapTotal(this.f).unsafeStep()
+
+    override fun unsafeRunAsyncTotal(cb: (Either<Throwable, A>) -> Unit) =
+            throw AssertionError("Unreachable")
+
+    override fun unsafeRunTimedTotal(limit: Duration): Option<A> =
+            throw AssertionError("Unreachable")
+}
+
+internal data class Async<out A>(val cont: ((Either<Throwable, A>) -> Unit) -> Unit) : IO<A>() {
+    override fun <B> map(f: (A) -> B): IO<B> =
+            mapDefault(this, f)
+
+    override fun <B> flatMapTotal(f: AndThen<A, IO<B>>): IO<B> =
+            BindAsync(cont, f)
+
+    override fun attempt(): IO<Either<Throwable, A>> =
+            BindAsync(cont, attemptValue())
+
+    override fun unsafeStep(): IO<A> =
+            this
+
+    override fun unsafeRunAsyncTotal(cb: (Either<Throwable, A>) -> Unit) =
+            cont(cb)
+
+    override fun unsafeRunTimedTotal(limit: Duration): Option<A> =
+            unsafeResync(this, limit)
+}
+
+internal data class BindAsync<E, out A>(val cont: ((Either<Throwable, E>) -> Unit) -> Unit, val f: AndThen<E, IO<A>>) : IO<A>() {
+    override fun <B> map(f: (A) -> B): IO<B> =
+            mapDefault(this, f)
+
+    override fun <B> flatMapTotal(ff: AndThen<A, IO<B>>): IO<B> =
+            BindAsync(cont, f.andThen(AndThen({ it.flatMapTotal(ff) }, { ff.error(it, { RaiseError(it) }) })))
+
+    override fun attempt(): IO<Either<Throwable, A>> =
+            BindSuspend(AndThen { _ -> this }, attemptValue())
+
+    override fun unsafeStep(): IO<A> =
+            this
+
+    override fun unsafeRunAsyncTotal(cb: (Either<Throwable, A>) -> Unit) =
+            cont { result ->
+                try {
+                    when (result) {
+                        is Either.Right -> f(result.b).unsafeRunAsync(cb)
+                        is Either.Left -> f.error(result.a, { RaiseError(it) }).unsafeRunAsync(cb)
+                    }
+                } catch (throwable: Throwable) {
+                    cb(Either.Left(throwable))
+                }
+            }
+
+    override fun unsafeRunTimedTotal(limit: Duration): Option<A> =
```

Comment:
I think returning `Option` in any of the unsafe methods will be confusing to users. We should return the expected value and since those methods are already unsafe let any exceptions propagate if unfolding the Option returned a NoSuchElementException.